sdks/python/apache_beam/typehints/native_type_compatibility.py

```diff
@@ -95,8 +101,15 @@ def _match_issubclass(match_against):
 
 def _match_same_type(match_against):
   # For Union types. They can't be compared with isinstance either, so we
-  # have to compare their types directly.
-  return lambda user_type: type(user_type) == type(match_against)
+  # Have to compare their types directly.
+
+  def matcher(derived, parent):
+    try:
+      return derived.__origin__ is parent
```

Comment:
Is there a unit test for this? I'm wondering why this code is not:
```suggestion
      return derived.__origin__ is parent.__origin__
```